sdformat_rs/src/lib.rs

```diff
@@ -11,10 +12,186 @@ use yaserde_derive::{YaDeserialize, YaSerialize};
 // Most of the structs are generated automatically from the
 include!(concat!(env!("OUT_DIR"), "/sdf.rs"));
 
+#[derive(Default, PartialEq, Clone, Debug)]
+pub enum ElementData {
+    #[default]
+    Empty,
+    Integer(i64),
+    Double(f64),
+    String(String),
+    Nested(HashMap<String, XmlElement>),
+}
+
+impl ElementData {
+    fn to_string(&self) -> String {
+        match self {
+            ElementData::Empty => "".into(),
+            ElementData::Integer(val) => val.to_string(),
+            ElementData::Double(val) => val.to_string(),
+            ElementData::String(val) => val.clone(),
+            ElementData::Nested(_) => unreachable!(),
+        }
+    }
+}
+
+#[derive(Default, PartialEq, Clone, Debug)]
+pub struct XmlElement {
+    pub attributes: HashMap<String, ElementData>,
+    pub data: ElementData,
+}
+
 // Manually declare plugin
-#[derive(Default, PartialEq, Clone, Debug, YaSerialize, YaDeserialize)]
-#[yaserde(rename = "plugin")]
-pub struct SdfPlugin {}
+#[derive(Default, PartialEq, Clone, Debug)]
+pub struct SdfPlugin {
+    pub name: String,
+    pub filename: String,
+    pub elements: HashMap<String, XmlElement>,
```

Comment:
Sorry, I missed this in the previous review. A hashmap string, limits us to 1 element of a given name. In some plugins, you can get repeated fields. Additionally, in those plugins order also matters. We might want a data structure `SdfElement` which contains a `HashMap<string, usize> and a `Vec<XmlElement>`.

Reply:
Nice catch, I totally didn't think of duplicated elements with the same name and we should definitely support that.
I agree with the ordering being significant, for example if a user was to specify a trajectory as a list of `<point>` elements, but are you aware of cases in which the order of heterogeneous elements is also significant? I wonder if it is rare enough that we could get away with a `HashMap<String, Vec<XmlElement>>`, it would also be a bit more friendly to work with, so if users have a list of repeated elements they can just do:

```
plugin.elements.get("point").and_then(|points| points.iter()).map(|p| {
[...]
});
```

Rather than

```
plugin.indexes.get("point").and_then(|indexes| indexes.iter().map(|i| plugin.elements.get(i)).map(|p| {
[...]
});
```

Reply:
`HashMap<String, Vec<XmlElement>>` would not work well for certain cases. For instance the buoyancy plugin in Gazebo does have heterogeneous element ordering that matters. The best we can do is encapsulate  `HashMap<string, HashSet<usize>>` with a `fn get(str: &str) -> Option<XmlElement>`, `fn push(str: &str, elem: XmlElement)` and an `fn iter()` method. I realize that this is a fair deal of work but would lead to a cleaner API. I'll have to give the API a bit more of thought to see how we can make construction easier as well.

Reply:
I tried to implement the suggestion in https://github.com/open-rmf/sdf_rust_experimental/pull/5/commits/1e30c6fe050748833199fc70cb967f05945bcf39.

Notable changes:
* Make the `get` API return references, clones could be expensive especially if users get an element that has a lot of nested data.
* Add an API `get_all` that returns a vector of references to help with cases of users wanting to get all the elements with a specific name.
* Store indexes in a `BTreeSet<usize>` instead of `HashSet<usize>` to make sure that the `get_all` API returns elements in order (useful for example for a series of points that make up a trajectory).
* Use `Arc<str>` instead of `String` for element name. I had to store the name in the `XmlElement` struct to make sure that the element name would be accessible to users iterating (previously it was only stored in the hash map as a key). For this reason there are two copies of the name present, the hash map and the vector, so I switched to `Arc<str>` to reduce memory allocations and reuse strings since they are immutable.
* Create an `all()` instead of `iter()`, this was really a bit out of lazyness since from what I understand implementing an iterator would need us to create yet a new struct that keeps the iterator state (iterator's `next()` take a `&mut self`) so that would have been a fair bit of boilerplate, I thought an `all()` function that just returns a slice that users can call `iter()` on would be reasonably simple.

Reply:
These all seem fairly reasonable to me.